src/github.com/getlantern/netx/copy.go

```diff
@@ -0,0 +1,72 @@
+package netx
+
+import (
+	"io"
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+var (
+	copyTimeout = 1 * time.Second
+)
+
+// BidiCopy copies between in and out in both directions, returning the errors
+// from copying to out and copying to in.
+func BidiCopy(out net.Conn, in net.Conn) (outErr error, inErr error) {
+	pool := getPool()
+	bufin := pool.Get()
+	bufout := pool.Get()
+	defer pool.Put(bufin)
+	defer pool.Put(bufout)
```

Comment:
Note - using a buffer pool is not necessary to fix this bug, but it's a useful optimization anyway.